jetty-util/src/main/java/org/eclipse/jetty/util/thread/ReservedThreadExecutor.java

```diff
@@ -336,63 +336,75 @@ private Runnable reservedWait()
 
-                if (_starting)
-                {
-                    if (LOG.isDebugEnabled())
-                        LOG.debug("{} started", this);
-                    _pending.decrementAndGet();
-                    _starting = false;
-                }
+                    // Are we surplus to capacity?
+                    if (size >= _capacity)
+                    {
+                        if (LOG.isDebugEnabled())
+                            LOG.debug("{} size {} >= capacity", this, size, _capacity);
+                        if (_starting)
+                            _pending.decrementAndGet();
+                        return;
+                    }
 
-                // Insert ourselves in the stack. Size is already incremented, but
-                // that only effects the decision to keep other threads reserved.
-                _stack.offerFirst(this);
+                    // If we cannot update size then recalculate
+                    if (!_size.compareAndSet(size, size + 1))
+                        continue;
 
-                // Once added to the stack, we must always wait for a job on the _task Queue
-                // and never return early, else we may leave a thread blocked offering a _task.
-                Runnable task = reservedWait();
+                    if (_starting)
+                    {
+                        if (LOG.isDebugEnabled())
+                            LOG.debug("{} started", this);
+                        _pending.decrementAndGet();
+                        _starting = false;
+                    }
 
-                if (task == STOP)
-                    // return on STOP poison pill
-                    break;
+                    // Insert ourselves in the stack. Size is already incremented, but
+                    // that only effects the decision to keep other threads reserved.
+                    _stack.offerFirst(this);
 
-                // Run the task
-                try
-                {
-                    task.run();
-                }
-                catch (Throwable e)
-                {
-                    LOG.warn(e);
+                    // Once added to the stack, we must always wait for a job on the _task Queue
+                    // and never return early, else we may leave a thread blocked offering a _task.
+                    Runnable task = reservedWait();
+
+                    if (task == STOP)
+                        // return on STOP poison pill
+                        break;
+
+                    // Run the task
+                    try
+                    {
+                        task.run();
+                    }
+                    catch (Throwable e)
+                    {
+                        LOG.warn(e);
+                    }
                 }
             }
-
-            if (LOG.isDebugEnabled())
-                LOG.debug("{} Exited", this);
+            catch (Throwable x)
+            {
+                LOG.warn(x);
```

Comment:
This doesn't do what it says in the PR description. If `reservedWait` throws, the reserved thread is left on the stack?

Reply:
You're correct. Yet another overlook.